packages/compass-data-modeling/src/store/relationships.ts

```diff
@@ -0,0 +1,215 @@
+ *                 values for _id field
+ *               * If the returned count equals the amount of sample values,
+ *                 return relationship
+ *
+ * @param foreignNamespace collection that is assumed "foreign" in the
+ * relationship
+ * @param foreignSchema schema of the "foreign" collection
+ * @param _sampleDocs
+ * @param collections list of all collections that will be checked for matching
+ * relationships
+ * @param dataService dataService instance
+ * @param abortSignal signal to cancel the inferring process
+ * @param onError callback that will be called if inference fails with an error
+ * @returns a list of confirmed relationships
+ */
+export async function inferForeignToLocalRelationshipsForCollection(
+  foreignNamespace: string,
+  foreignSchema: MongoDBJSONSchema,
+  _sampleDocs: Document[],
+  collections: { ns: string; schema: MongoDBJSONSchema; sample: Document[] }[],
+  dataService: DataService,
+  abortSignal?: AbortSignal,
+  onError?: (err: any) => void
+): Promise<Relationship['relationship'][]> {
+  const idSchema = foreignSchema.properties?._id;
+  if (!idSchema) {
+    return [];
+  }
+  const indexes = await dataService
+    .indexes(foreignNamespace, { full: false })
+    .catch(() => {
+      return [];
+    });
+  const hasIdIndex = indexes.some((definition) => {
+    return (
+      definition.fields.length === 1 && definition.fields[0].field === '_id'
+    );
+  });
+  if (!hasIdIndex) {
+    return [];
+  }
+  const relationships = await Promise.all(
+    collections.flatMap((localColl) => {
+      if (localColl.ns === foreignNamespace) {
+        return [];
+      }
+      const schemaPaths = findPropertyPathsMatchingSchema(
+        localColl.schema,
+        idSchema
+      ).filter((value) => {
+        // They will be matching in a lot of cases, but the chances of both
+        // local and foreign field in a relationship being _id are very slim, so
+        // skipping
+        return value[0] !== '_id';
+      });
+      return schemaPaths.map(
+        async (propPath): Promise<Relationship['relationship'] | null> => {
+          try {
+            const sampleDocs = localColl.sample
+              .flatMap((doc) => {
+                return getValuesFromPath(doc, propPath);
+              })
+              .filter((doc) => {
+                // remove missing values from the data sample
+                return doc !== undefined && doc !== null;
+              })
+              // in case sample data is an array that contains a lot of values,
+              // we limit the amount of samples to reduce the matching time
+              .slice(0, 100);
+            if (sampleDocs.length === 0) {
+              return null;
+            }
+            const matchingDocCount = await dataService.count(
+              foreignNamespace,
+              {
+                _id: {
+                  $in: sampleDocs as any[], // driver wants this to be an ObjectId?
```

Comment:
Yeah, it's trying to be helpful by assuming that `_id` is always an `ObjectId` unless you explicitly use something like `db.collection<{ _id: any }>('foo').count(...)`. It feels a bit overeager to me too but honestly for the typical application developer it's probably more helpful than unhelpful

Reply:
Hmm, seems like specifically passing `any` as a type for a field is tripping up the generic logic for the Filter type in the driver ending up requiring `_id` to be `never`. I'll leave this as an assertion for now, but will revisit